lib/multi_channel.ml

```diff
@@ -0,0 +1,180 @@
+  { mask = sz - 1;
+    channels = Array.init sz (fun _ -> Ws_deque.create ());
+    waiters = Chan.make_unbounded ();
+    next_domain_id = Atomic.make 0;
+    recv_block_spins;
+    }
+
+let register_domain mchan =
+  let id = Atomic.fetch_and_add mchan.next_domain_id 1 in
+  assert(id < Array.length mchan.channels);
+  id
+
+let init_domain_state mchan dls_state =
+  let id = (register_domain mchan) in
+  dls_state.id <- id;
+  dls_state.steal_offsets <- Array.init ((Array.length mchan.channels)-1) (fun i -> i+1);
+  dls_state
+  [@@inline never]
+
+let get_local_id mchan =
+  let dls_state = Domain.DLS.get dls_key in
+  if dls_state.id >= 0 then dls_state.id
+  else (init_domain_state mchan dls_state).id
+  [@@inline]
+
+let clear_local_state () =
+  let dls_state = Domain.DLS.get dls_key in
+  dls_state.id <- (-1)
+
+let rec check_waiters mchan =
+  Domain.Sync.poll (); (* need to make sure we have a safepoint in here *)
+  match Chan.recv_poll mchan.waiters with
+    | None -> ()
+    | Some (status, mc) ->
+      (* avoid the lock if we possibly can *)
+      if !status = Released then check_waiters mchan
+      else begin
+        Mutex.lock mc.mutex;
+        match !status with
+        | Waiting ->
+          begin
+            status := Released;
+            Mutex.unlock mc.mutex;
+            Condition.broadcast mc.condition
+          end
+        | Released ->
+          begin
+            (* this waiter is already released, it might have found something on a poll *)
+            Mutex.unlock mc.mutex;
+            check_waiters mchan
+          end
+      end
+
+let send mchan v =
+  let id = (get_local_id mchan) in
+  Ws_deque.push (Array.unsafe_get mchan.channels id) v;
+  check_waiters mchan
+
+let rec recv_poll_loop mchan dls cur_offset =
+  Domain.Sync.poll (); (* need to make sure we have a safepoint in here *)
+  let offsets = dls.steal_offsets in
+  let k = (Array.length offsets) - cur_offset in
+  if k = 0 then raise Exit
+  else begin
+    let idx = cur_offset + (Random.State.int dls.rng_state k) in
+    let t = Array.unsafe_get offsets idx in
+    let channel = Array.unsafe_get mchan.channels (Int.logand (dls.id + t) mchan.mask) in
+    try
+      Ws_deque.steal channel
+    with
+      | Exit ->
+        begin
+          Array.unsafe_set offsets idx (Array.unsafe_get offsets cur_offset);
+          Array.unsafe_set offsets cur_offset t;
+          recv_poll_loop mchan dls (cur_offset+1)
+        end
+  end
+
+let recv_poll mchan =
+  Domain.Sync.poll (); (* need to make sure we have a safepoint in here *)
+  let id = (get_local_id mchan) in
+  try
+    Ws_deque.pop (Array.unsafe_get mchan.channels id)
+  with
+    | Exit -> recv_poll_loop mchan (Domain.DLS.get dls_key) 0
+
+let rec recv_poll_repeated mchan repeats =
+  try
+    recv_poll mchan
+  with
+    | Exit ->
+      if repeats = 1 then raise Exit
+      else begin
+        Domain.Sync.cpu_relax ();
+        recv_poll_repeated mchan (repeats - 1)
```

Comment:
Wondering if it'd be worth removing repeated calls to `Domain.DLS.get` in this path. This may not be actually important though, because the calling domain is likely to block later on.

Reply:
You are right, we can reduce the number of calls to `Domain.DLS.get`. I've restructured things to do this.